lib/projfs.c

```diff
@@ -624,50 +663,90 @@ static int project_dir(const char *op, const char *path, int parent)
 static int project_file(const char *op, const char *path,
 			enum proj_state state)
 {
+	char self_fd_path[MAX_PROC_SELF_FD_PATH_LEN + 1];
 	struct proj_state_lock state_lock;
+	struct stat st;
+	int reset_mode = 0;
 	int log = 0;
-	int res;
+	int lock_fd;
+	int fd, res;
+
+	if (state == PROJ_STATE_EMPTY)
+		return EINVAL;
 
 	/* Pass O_NOFOLLOW so we receive ELOOP if path is an existing symlink,
-	 * which we want to ignore, and request a write mode so we receive
-	 * EISDIR if path is a directory.
+	 * which we want to ignore.
 	 */
-	res = acquire_proj_state_lock(&state_lock, path, O_RDWR | O_NOFOLLOW);
+	res = acquire_proj_state_lock(&state_lock, path,
+				      O_RDONLY | O_NOFOLLOW | O_NONBLOCK);
 	if (res != 0) {
 		if (res == ELOOP)
 			return 0;
 		else
 			return res;
 	}
 
-	// hydrate empty placeholder file
-	if (state_lock.state == PROJ_STATE_EMPTY) {
-		struct stat st;
-		int reset_mtime;
+	lock_fd = state_lock.lock_fd;
+	if (fstat(lock_fd, &st) == -1) {
+		res = errno;
+		goto out_release;
+	}
+	else if (!S_ISREG(st.st_mode)) {
+		if (S_ISDIR(st.st_mode))
+			res = EISDIR;
+		else
+			res = 0;
+		goto out_release;
+	}
 
-		reset_mtime = (fstat(state_lock.lock_fd, &st) == 0);
+	// check after fstat() because we need to return EISDIR if not a file
+	if (state_lock.state == state)
+		goto out_release;
+
+	// TODO: for non-Linux, may need to use other technique to reopen file
+	sprintf(self_fd_path, PROC_SELF_FD_PATH_FMT, lock_fd);
+	fd = open(self_fd_path, O_WRONLY | O_NONBLOCK);
+	if (fd == -1) {
+		res = errno;
+		reset_mode = fchmod_user_write_stat(lock_fd, &st, 1);
+		if (!reset_mode)
+			goto out_release;
+		fd = open(self_fd_path, O_WRONLY | O_NONBLOCK);
+		if (fd == -1) {
+			res = errno;
+			goto out_release;
+		}
+	}
```

Comment:
`res` will carry the `errno` value (set on line 710) all the way to the end of the function without being reset. This leads to the `EACCES` from our failed `open` on 708 being propagated to the caller (despite successfully adding and removing the `IWUSR` bit), as seen in this example using `test_simple`:

```console
$ ls -la ~/mount/
total 12
drwxr-xr-x  2 kivikakk kivikakk 4096 Jun  5 14:20 ./
drwxr-xr-x 26 kivikakk kivikakk 4096 Jun  5 14:17 ../
-r--r--r--  1 kivikakk kivikakk    6 May 21 12:25 hi
$ cat ~/mount/hi
cat: /home/kivikakk/mount/hi: Permission denied
$
```

We may want this:

```suggestion
		res = 0;
	}
```

I found this because curiosity provoked me trying this, which still fails on this branch:

```diff
diff --git a/t/t000-mirror-read.t b/t/t000-mirror-read.t
index bbdbcd7..078396b 100755
--- a/t/t000-mirror-read.t
+++ b/t/t000-mirror-read.t
@@ -33,7 +33,8 @@ test_expect_success 'create source tree' '
        echo file1 > source/f1.txt &&
        echo file2 > source/f2.txt &&
        echo file1 > source/d1/f1.txt &&
-       echo file2 > source/d1/d2/f2.txt
+       echo file2 > source/d1/d2/f2.txt &&
+       chmod 444 source/f1.txt
 '

 test_expect_success 'check target tree structure' '
```

With the above suggestion, the test passes. We probably want to add some dedicated tests for this case, as well as a FIFO test.

Reply:
Great catch, thank you!  👍

Reply:
Also, looking closely, there's an oversight a bit earlier too -- this should have a `>=`:
```
if (state_lock.state == state)
```
While writing the test scripts, I began to wonder, why is this non-projected file even trying to be opened for writing here?  Sigh.

Reply:
I think using any operator other than equality/inequality on enums is an antipattern -- what if we add another value that semantically makes no sense here? It's a bit awkward given the RHS is an argument though ... _scratches head_

Reply:
Switched to a more explicit `state_lock.state == state || state_lock.state == PROJ_STATE_MODIFIED`.